### PR TITLE
Avoid activity task prefetching when handlers are busy

### PR DESCRIPTION
**Pull request by vitarb**
Note, This PR is heavily based on #278 proposal by Maxim, which superseded initial implementation in #277.

On top of that I've addressed an issue with late semaphore acquisition and added a test that verifies that poller behavior is correct and tasks are not getting pre-fetched when handlers are busy.

As a bonus I've created a WorkerRule, which takes care of the common boilerplate used in tests to create workflow workers and can be used to write tests easily without needing to expand already gigantic WorkflowTest.

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import io.temporal.api.workflowservice.v1.PollActivityTaskQueueResponse;
+import io.temporal.workflow.Functions;
+
+public class ActivityTask {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java:25]: Let's change from "public class" to "final class"

Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java]: done

File: temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  public ActivityTask(PollActivityTaskQueueResponse response, Functions.Proc completionHandle) {
+    this.response = response;
+    this.completionHandle = completionHandle;
+  }
+
+  public PollActivityTaskQueueResponse getResponse() {
+    return response;
+  }
+
+  public Functions.Proc getCompletionHandle() {
```
Comment (Sushisource) [inline temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java:38]: Might be worth adding a docstring indicating that owners are expected to call this when the task is complete

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java [hunk 3]
```diff
@@ -114,6 +143,14 @@ public Scope getMetricsScope() {
     return metricsScope;
   }
 
+  public boolean isUseExternalService() {
```
Comment (Sushisource) [inline temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java:146]: Docstring for what this means would be useful as well

File: temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import io.temporal.testing.TestWorkflowEnvironment;
+import io.temporal.worker.Worker;
+import io.temporal.worker.WorkerFactoryOptions;
+import io.temporal.worker.WorkerOptions;
+import java.util.UUID;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public class WorkerRule implements TestRule {
```
Comment (Sushisource) [inline temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java:34]: Are there other existing tests that could be shortened by using this?

Comment (vitarb) [reply temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java:34]: All of them.

Comment (mfateev) [inline temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java:34]: I would rename it to TestWorkflowRule to be consistent with TestWorkflowEnvironment

File: temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java (hunk 2 above)
Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityTask.java:38]: Done.

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java (hunk 3 above)
Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java:146]: Done

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java [hunk 5]
```diff
@@ -54,11 +54,17 @@ public static TestEnvironmentOptions getDefaultInstance() {
 
     private Scope metricsScope;
 
+    private boolean useExternalService;
+
+    private String serviceAddress;
```
[the remark this answers is not in the document]
Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java:59]: Done

File: temporal-sdk/src/test/java/io/temporal/workflow/ActivityPollerPrefetchingTest.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ActivityPollerPrefetchingTest {
+
+  @Rule
+  public WorkerRule workerRule =
+      new WorkerRule.Builder()
```
[the remark this answers is not in the document]
Comment (vitarb) [reply temporal-sdk/src/test/java/io/temporal/workflow/ActivityPollerPrefetchingTest.java:40]: Done

File: temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+  private Class<?>[] workflowTypes;
+  private Object[] activityImplementations;
+  private TestWorkflowEnvironment testEnvironment;
+  private WorkerOptions workerOptions;
+  private Worker worker;
+  private String taskQueue;
+  private boolean useExternalService;
```
[the remark this answers is not in the document]
Comment (vitarb) [reply temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java:42]: We can't do them all final as worker/task queue are initialized in the init method that is called from apply rather then constructor, but most of them can.

File: temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java (hunk 4 above)
Comment (vitarb) [reply temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java:34]: Done

File: temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      Object[] activityImplementations,
+      WorkerOptions workerOptions) {
+    this.testEnvironment = testEnvironment;
+    this.useExternalService = useExternalService;
+    this.workflowTypes = workflowTypes;
+    this.activityImplementations = activityImplementations;
+    this.workerOptions = workerOptions;
+  }
+
+  static class Builder {
```
[the remark this answers is not in the document]
Comment (vitarb) [reply temporal-sdk/src/test/java/io/temporal/workflow/WorkerRule.java:57]: Done

**Review by Sushisource — APPROVED** [review]

File: temporal-sdk/src/test/java/io/temporal/workflow/TestWorkflowRule.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+      this.activityImplementations = activityImplementations;
+      return this;
+    }
+
+    public Builder setUseExternalService(boolean useExternalService) {
+      this.useExternalService = useExternalService;
+      return this;
+    }
+
+    public Builder setTarget(String target) {
```
Comment (mfateev) [inline temporal-sdk/src/test/java/io/temporal/workflow/TestWorkflowRule.java:96]: Add JavaDoc explaining the meaning of the target and useExternalService properties.

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java [hunk 10]
```diff
@@ -77,29 +83,52 @@ public Builder setMetricsScope(Scope metricsScope) {
       return this;
     }
 
+    public Builder setUseExternalService(boolean useExternalService) {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java:86]: rename to target and add appropriate JavaDoc.

File: temporal-sdk/src/test/java/io/temporal/workflow/TestWorkflowRule.java (hunk 9 above)
Comment (vitarb) [reply temporal-sdk/src/test/java/io/temporal/workflow/TestWorkflowRule.java:96]: Added here as well  as also included a java doc with example for the entire rule.

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java (hunk 10 above)
Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java:86]: Do you want to rename to `setUseExternalTarget`? Added a javadoc.

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java [hunk 11]
```diff
@@ -77,29 +83,52 @@ public Builder setMetricsScope(Scope metricsScope) {
...
+      this.useExternalService = useExternalService;
+      return this;
+    }
+
+    public Builder setServiceAddress(String serviceAddress) {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java]: Rename to setTarget

File: temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java (hunk 10 above)
Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/testing/TestEnvironmentOptions.java:86]: Addressed the naming in `WorkerFactoryOptions`

File: temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityWorker.java [hunk 12]
```diff
@@ -161,44 +160,44 @@ private TaskHandlerImpl(ActivityTaskHandler handler) {
...
-      MDC.put(LoggerTag.RUN_ID, task.getWorkflowExecution().getRunId());
+      MDC.put(LoggerTag.ACTIVITY_ID, r.getActivityId());
+      MDC.put(LoggerTag.ACTIVITY_TYPE, r.getActivityType().getName());
+      MDC.put(LoggerTag.WORKFLOW_ID, r.getWorkflowExecution().getWorkflowId());
+      MDC.put(LoggerTag.RUN_ID, r.getWorkflowExecution().getRunId());
 
-      propagateContext(task);
+      propagateContext(r);
 
       try {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityWorker.java:189]: Let's move the try to the beginning of the method. For example, if `propagateContext` throws we are going to leak semaphore.

Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityWorker.java]: Done, we still need to have metric scope and response variables declared outside of the try block though.
I think it should be safe this way.

File: temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityPollTask.java [hunk 13]
```diff
@@ -86,31 +89,39 @@ public PollActivityTaskQueueResponse poll() {
...
-      result =
-          service
-              .blockingStub()
-              .withOption(METRICS_TAGS_CALL_OPTIONS_KEY, metricsScope)
-              .pollActivityTaskQueue(pollRequest.build());
-    } catch (StatusRuntimeException e) {
-      if (e.getStatus().getCode() == Status.Code.UNAVAILABLE
-          && e.getMessage().startsWith("UNAVAILABLE: Channel shutdown")) {
-        return null;
+      pollSemaphore.acquire();
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityPollTask.java:95]: I believe we have to move this acquire to before the try

Comment (vitarb) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityPollTask.java:96]: done

**Review by mfateev — APPROVED** [review]

**Pull request merged**